### ISSUE TITLE
openforcefield fails when OEMol misses atom names

### ISSUE THREAD

**Issue by nividic** [open-forcefield-group/openforcefield#47]
I was trying to parametrize an OEMol by using the SMIRNOFF FF. However the parametrization fails with errors:

Traceback (most recent call last):
  File "test.py", line 20, in <module>
    mol_top, mol_sys, mol_pos = create_system_from_molecule(mol_ff, ligand)
  File "/Users/gcalabro/local/miniconda2/envs/openeye/lib/python3.5/site-packages/smarty/forcefield_utils.py", line 101, in create_system_from_molecule
    system = forcefield.createSystem(topology, [mol], verbose=verbose)
  File "/Users/gcalabro/local/miniconda2/envs/openeye/lib/python3.5/site-packages/smarty/forcefield.py", line 950, in createSystem
    topology = _Topology(topology, molecules)
  File "/Users/gcalabro/local/miniconda2/envs/openeye/lib/python3.5/site-packages/smarty/forcefield.py", line 230, in __init__
    self._identifyMolecules()
  File "/Users/gcalabro/local/miniconda2/envs/openeye/lib/python3.5/site-packages/smarty/forcefield.py", line 361, in _identifyMolecules
    raise Exception(msg)
Exception: No provided molecule matches topology molecule:
Atom        0           0 pBace_lLigand 1_0

I was able to narrow down the error and it is related to missing string atom names in the passed OEMol. In particular the dictionary in the function [generateTopologyFromOEMol](https://github.com/open-forcefield-group/openforcefield/blob/master/openforcefield/utils/utils.py#L186) is generated with just one entry. To overcome the problem I called the function:
```python
oechem.OETriposAtomNames(molecule)
```
prior to parametrize my molecule and the problem was fixed. I advice to add a check to the function [generateTopologyFromOEMol](https://github.com/open-forcefield-group/openforcefield/blob/master/openforcefield/utils/utils.py#L157) to test if any of the string atom name is empty and if so assign Tripos atom names:
```python
    if any([atom.GetName() == '' for atom in molecule.GetAtoms()]):
        oechem.OETriposAtomNames(molecule)
```

**Comment by davidlmobley** [comment]
Thanks, @nividic . That sounds like an excellent plan.

Probably at the same time I should update this functionality to add bond orders to the OpenMM `Topology` if they are present in the OEMol, as this is now supported. You have code which does this on https://github.com/oess/openmm_orion, right? Can you point me to the relevant functionality? This would ensure that if you ever have to turn these `Topology` objects back into `OEMol`s it would work properly.

**Comment by davidlmobley** [comment]
CC #https://github.com/oess/openmm_orion/pull/62

**Comment by nividic** [comment]
These two functions are what you are looking for:

[oemol_to_openmmTop](https://github.com/oess/openmm_orion/blob/master/ComplexPrepCubes/utils.py#L126)

and

[openmmTop_to_oemol](https://github.com/oess/openmm_orion/blob/master/ComplexPrepCubes/utils.py#L192)

**Comment by nividic** [comment]
I have just found out that also if you have not empty atom name strings the parametrization can still fail. The string atom name MUST be Tripos Atom name and therefore you have to call:
```python 
 oechem.OETriposAtomNames(molecule)
```

**Comment by davidlmobley** [comment]
@nividic - it's not immediately clear to me why these would have to be Tripos atom names, as opposed to any unique string. Can you clarify/provide an example to reproduce?

**Comment by davidlmobley** [comment]
(Though, if we're converting a Topology to an OEMol internally the naming convention won't be important for us to retain, so it should be OK for us to just always use Tripos names. I'm just trying to think through why this would be a requirement.)

**Comment by nividic** [comment]
May be I'm wrong...I was checking a TLA excipient for parametrization with Smirnoff ff and it was failing. I checked that it had associated string atom names but I did not payed attention if the atom names were unique

**Comment by davidlmobley** [comment]
Closed via #49
